app/code/Magento/InventoryShipping/Observer/VirtualSourceDeductionProcessor.php

```diff
@@ -0,0 +1,271 @@
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->getSourceItemBySourceCodeAndSku = $getSourceItemBySourceCodeAndSku;
+        $this->getSkusByProductIds = $getSkusByProductIds;
+        $this->getStockItemConfiguration = $getStockItemConfiguration;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->inventoryRequestFactory = $inventoryRequestFactory;
+        $this->itemRequestFactory = $itemRequestFactory;
+        $this->sourceSelectionService = $sourceSelectionService;
+        $this->getDefaultSourceSelectionAlgorithmCode = $getDefaultSourceSelectionAlgorithmCode;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Invoice $invoice */
+        $invoice = $observer->getEvent()->getInvoice();
+
+        if ($invoice->getOrigData('entity_id')) {
+            return;
+        }
+
+        $selectionRequestItems = [];
+        foreach ($invoice->getItems() as $invoiceItem) {
+            if (!$this->isValidItem($invoiceItem)) {
+                continue;
+            }
+
+            $itemSku = $invoiceItem->getSku() ?: $this->getSkusByProductIds->execute(
+                [$invoiceItem->getProductId()]
+            )[$invoiceItem->getProductId()];
+            $qty = $this->castQty($invoiceItem->getOrderItem(), $invoiceItem->getQty());
+
+            $selectionRequestItems[] = $this->itemRequestFactory->create([
+                'sku' => $itemSku,
+                'qty' => $qty,
+            ]);
+        }
+
+        if (empty($selectionRequestItems)) {
+            return;
+        }
+
+        $order = $invoice->getOrder();
+        $websiteId = $order->getStore()->getWebsiteId();
+        $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
+
+        /** @var InventoryRequestInterface $inventoryRequest */
+        $inventoryRequest = $this->inventoryRequestFactory->create([
+            'stockId' => $stockId,
+            'items' => $selectionRequestItems
+        ]);
+        $sourceSelectionResult = $this->sourceSelectionService->execute(
+            $inventoryRequest,
+            $this->getDefaultSourceSelectionAlgorithmCode->execute()
+        );
+
+        $this->deductSources($sourceSelectionResult, $stockId);
+    }
+
+    /**
+     * @param InvoiceItemInterface $invoiceItem
+     * @return bool
+     */
+    private function isValidItem(InvoiceItemInterface $invoiceItem): bool
+    {
+        $orderItem = $invoiceItem->getOrderItem();
+        return in_array(
+            $orderItem->getProductType(),
+            [
+                VirtualType::TYPE_VIRTUAL,
+                DownloadableType::TYPE_DOWNLOADABLE
+            ]
```

Comment:
It's better to remove high coupling on ProductTypes from the code and pass additional variable `$productTypesToDeductOnInvoice` as a constructor argument. And adjust this argument via DI settings

```
    <type name="Magento\InventoryShipping\Observer\VirtualSourceDeductionProcessor">
        <arguments>
            <argument name="productTypesToDeductOnInvoice" xsi:type="array">
                <item name="virtual" xsi:type="const">Magento\Catalog\Model\Product\Type::TYPE_VIRTUAL</item>
                <item name="downloadable" xsi:type="const">Magento\Downloadable\Model\Product\Type::TYPE_DOWNLOADABLE</item>
            </argument>
        </arguments>
   </type>
```

doing so, ProductTypes become an extension point and it would be possible easily to extend ProductType list

Reply:
I use 
```php
if (!$invoiceItem->getOrderItem()->getIsVirtual()) {
     continue;
}
```
instead... probably this need to changed, but then we should change `\Magento\InventorySourceSelection\Model\InventoryRequestFromOrderFactory::create line 81` in the same way